src/core/remio.c

```diff
@@ -0,0 +1,641 @@
+            ERROR("Invalid shared memory region configuration for Remote I/O device %d.\n"
+                  "The frontend and backend shared memory regions must be the aligned.",
+                dev->bind_key);
+        }
+    }
+
+    /** Update the Remote I/O device configuration */
+    for (size_t vm_id = 0; vm_id < config.vmlist_size; vm_id++) {
+        struct vm_config* vm_config = &config.vmlist[vm_id];
+        for (size_t i = 0; i < vm_config->platform.remio_dev_num; i++) {
+            struct remio_dev* dev = &vm_config->platform.remio_devs[i];
+            struct remio_device* device = remio_find_dev_by_bind_key(dev->bind_key);
+            if (device == NULL) {
+                ERROR("Failed to find Remote I/O device %d", dev->bind_key);
+            }
+            if (dev->type == REMIO_DEV_BACKEND) {
+                device->config.backend.vm_id = vm_id;
+                device->config.backend.interrupt = dev->interrupt;
+                device->config.backend.cpu_id = (cpuid_t)-1;
+            } else if (dev->type == REMIO_DEV_FRONTEND) {
+                device->config.frontend.vm_id = vm_id;
+                device->config.frontend.interrupt = dev->interrupt;
+                device->config.frontend.cpu_id = (cpuid_t)-1;
+            } else {
+                ERROR("Unknown Remote I/O device type");
+            }
+        }
+    }
+}
+
+void remio_assign_vm_cpus(struct vm* vm)
+{
+    list_foreach (remio_device_list, struct remio_device, dev) {
+        if (vm->id == dev->config.backend.vm_id) {
+            dev->config.backend.cpu_id = min(dev->config.backend.cpu_id, cpu()->id);
+            dev->config.backend.ready = true;
+        } else if (vm->id == dev->config.frontend.vm_id) {
+            dev->config.frontend.cpu_id = min(dev->config.frontend.cpu_id, cpu()->id);
+            dev->config.frontend.ready = true;
+        }
+        dev->ready = dev->config.backend.ready && dev->config.frontend.ready;
+    }
+}
+
+/**
+ * @brief Handles the Remote I/O ask operation
+ * @param addr Should always be zero (convention)
+ * @param value Should always be zero (convention)
+ * @param device Pointer to the Remote I/O device
+ * @return Returns the number of pending I/O requests
+ */
+static long int remio_handle_ask(unsigned long addr, unsigned long value,
+    struct remio_device* device)
+{
+    long int ret = -HC_E_SUCCESS;
+    long unsigned int pending_requests = 0;
+    struct remio_request* request = NULL;
+
+    /** By convention, the addr and value fields must be zero */
+    if (addr != 0 || value != 0) {
+        return HC_E_FAILURE;
+    }
+
+    /** Get the next pending I/O request */
+    request = remio_get_request(device, REMIO_NEXT_PENDING_REQUEST);
+
+    if (request == NULL || request->state != REMIO_STATE_PENDING) {
+        return HC_E_FAILURE;
+    }
+
+    /** Calculate the remaining number of pending I/O requests */
+    pending_requests = (long unsigned int)remio_num_pending_request(device);
```

Comment:
Is this really necessary? Can't we just return some flag that says that there are more requests waiting? If so we don't need to go through the list to count its size in list_size. Just use list_emtpy...

Reply:
I think it was the right way to go to remove it.